cp: Do not dereference symlink even when dangling - fix issue #3364

anastygnome opened the pull request:
Fixes an issue with cp not copying symlinks in spite of the -P  (no dereference option) 
fixes #3364
```console
> ln -s no-such-file dangle
> LANG=C cp dangle d2
cp: cannot stat 'dangle': No such file or directory
[1]> LANG=C cp -P dangle d2
> rm d2
>  ./uu_cp dangle d2
cp: cannot stat 'dangle': No such file or directory
[1]> ./uu_cp -P dangle d2
> rm d2
```

anastygnome commented on src/uu/cp/src/cp.rs:
```diff
@@ -1286,10 +1286,13 @@ fn copy_file(
+    // Fail if upon dereferencing symlink,
+    // dest a dangling symlink or a symlink this program created previously.
+
     if fs::symlink_metadata(dest)
         .map(|m| m.file_type().is_symlink())
         .unwrap_or(false)
+        && !options.dereference
```
maybe swapping the two clauses of the and test should be better to short-circuit the fs call.

tertsdiepraam commented on tests/by-util/test_cp.rs:
```diff
@@ -1496,14 +1496,29 @@ fn test_copy_through_just_created_symlink() {
+}
+
+#[test]
+fn test_copy_through_dangling_symlink_dereference() {
```
Shouldn't the name here be "no_dereference"?

anastygnome replied:
yup, will fix

tertsdiepraam reviewed (commented):
I think your test cases are backwards from your example in the PR. I.e. you do
```bash
cp (-P) link file # in the example
cp (-P) file link # in the test
```
It's probably best to test for both and check what the expected behaviour should be.

tertsdiepraam commented:
Did you remove the tests from the PR? I thought they were really close to being correct and it would be great to have them!

tertsdiepraam commented:
Don't worry about the failing Windows tests, I'm trying to fix those here: https://github.com/uutils/coreutils/pull/3460

anastygnome commented:
@tertsdiepraam I had forgot to add the tests. There are now back in the PR.

anastygnome commented:
I'm cleaning UP history, please stop for now

anastygnome commented:
@tertsdiepraam  @sylvestre 
please resume the workflow

sylvestre commented:
please rename the commit for something more explicit

anastygnome commented:
will do

anastygnome commented:
@sylvestre

anastygnome commented:
sorry for the hassle @sylvestre my PC is driving me nuts. 

Just caught a nasty bug.

sylvestre commented:
yeah, please push when it is ready ;)

anastygnome commented:
@sylvestre I'm done

anastygnome commented:
Can we enable feature is symlink or do you want me to revert this part of the commit @sylvestre

sylvestre commented:
For now, we haven't discussed about bumping MinRustV.

The error for others:
`error[E0658]: use of unstable library feature 'is_symlink'`

anastygnome commented:
I've brung back the old method using metadata for now and added a comment.

anastygnome commented:
rebased to pass tests  from main @sylvestre

anastygnome commented:
@sylvestre failing test is unrelated, good to be merged

sylvestre commented on tests/by-util/test_cp.rs:
```diff
@@ -1505,6 +1505,18 @@ fn test_copy_through_dangling_symlink() {
+#[test]
+fn test_copy_through_dangling_symlink_no_dereference() {
+    let (at, mut ucmd) = at_and_ucmd!();
+    at.touch("file");
+    at.symlink_file("nonexistent", "target");
+    ucmd.arg("file")
+        .arg("-P")
+        .arg("target")
```
According to https://github.com/uutils/coreutils/issues/3364#issue-1191201385=
it should be a bit different (and renamed to facilitate the match)

```suggestion
    at.symlink_file("no-such-file", "dangle");
    ucmd.arg("-P")
        .arg("dangle")
        .arg("d2")
```

sylvestre commented on tests/by-util/test_cp.rs:
```diff
@@ -1505,6 +1505,17 @@ fn test_copy_through_dangling_symlink() {
         .fails()
         .stderr_only("cp: not writing through dangling symlink 'target'");
 }
+#[test]
```
```suggestion
}

#[test]
```

sylvestre commented:
why did you push over my changes ?

anastygnome commented:
@sylvestre thank the colleague who put force by default in my pull and push parameters. 
fixed abd brought back yours. They had been squashed.

I just squashed the add space commit to reduce noised

anastygnome commented:
failing test has been corrected

sylvestre commented:
thanks :)
